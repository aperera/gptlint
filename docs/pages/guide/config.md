# Config

TODO

## Config File

TODO

```js
// gptlint.config.js

/** @type {import('gptlint').GPTLintConfig} */
export default [
  {
    llmOptions: {
      model: 'gpt-4-turbo-preview',
      weakModel: 'gpt-3.5-turbo'
    }
  }
]
```

```js
// gptlint.config.js

/** @type {import('gptlint').GPTLintConfig} */
export default [
  {
    llmOptions: {
      apiBaseUrl: 'https://openrouter.ai/api/v1',
      model: 'anthropic/claude-3-opus:beta',
      weakModel: 'anthropic/claude-3-haiku:beta',
      // Optional
      kyOptions: {
        headers: {
          // Optional, for including your app on openrouter.ai rankings.
          'HTTP-Referer': 'https://github.com/gptlint/gptlint',
          // Optional, shows in rankings on openrouter.ai.
          'X-Title': 'gptlint'
        }
      }
    }
  }
]
```
