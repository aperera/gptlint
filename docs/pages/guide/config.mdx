import { Callout } from 'nextra/components'

# Config

GPTLint config is inspired by [ESLint's new flat config format](https://eslint.org/docs/latest/use/configure/configuration-files), so if you've ever used `eslint`, then `gptlint` should feel very familiar.

Just like ESLint, GPTLint supports a variety of ways to customize and extend it's behavior and default linting rules. Configs can come from the following places, with more specific configs overriding more general settings:

- default settings
- project-specific config file (`gptlint.config.{js,mjs,cjs}`)
- CLI flags (all of the main config settings have corresponding CLI flags)
- inline config overrides in source files (`/* gptlint-disable */`)

In the near future, we'll be moving the current set of default JS/TS rules into an NPM package `@gptlint/typescript-config`, and in general this is the expected way to provide reusable sets of pre-configured rules going fowards. As a general principle, we're trying to follow ESLint's config design as closely as possible.

## Print Config

You can always run `gptlint --print-config` to print out exactly what rules, files, and settings are being resolved by GPTLint.

**The `--print-config` option is extremely useful** for making sure your project is set up correctly and for validating your understanding of how the linter works under the hood.

## Default Config

The default config is available like so:

```ts
import { recommendedConfig } from 'gptlint'
```

`recommendedConfig` is an array of `LinterConfig` objects which enable the [default rule set](/rules). If you do not specify a custom config file, then the recommended config will be applied by default. If you do specify a custom config file, then you must manually opt-in to the default rules like in the examples below. This design makes it easy to add support for third-party NPM packages and reusable bundles of custom configs instead of treating the built-in rules specially.

## Config File

The GPTLint configuration file may be named any of the following:

- `gptlint.config.js`
- `gptlint.config.mjs`
- `gptlint.config.cjs`

It should be placed in the root directory of your project and export an array of [configuration objects](#linterconfig-objects). Hereâ€™s an example:

```js filename="gptlint.config.js"
import { recommendedConfig } from 'gptlint'

export default [
  ...recommendedConfig,
  {
    rules: {
      'always-handle-promises': 'warn',
      'semantic-variable-names': 'off'
    }
  }
]
```

In this example, the configuration array contains just one configuration object. The configuration object will be merged with the default config to set `always-handle-promises` violations to warnings and to disable the `semantic-variable-names` rule for all files.

### LinterConfig Objects

LinterConfig objects are validated against a [LinterConfigSchema](https://github.com/gptlint/gptlint/tree/main/src/config.ts) zod schema which has a very similar shape to [ESLint Configuration Objects](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-objects).

```ts
export type LinterConfig = {
  files?: string[] // array of globs to search for source files
  ignores?: [] // array of globs to ignore for source files

  ruleFiles?: [] // array of globs to find project-specific rule md files
  ruleDefinitions?: RuleDefinition[] // custom rules (extensible to non-md formats)

  // enable / disable rules
  rules?: Record<string, 'error' | 'warn' | 'off'>

  // customize linter options (debug, cache dir, logging, etc)
  linterOptions?: LinterOptions

  // customize LLM model and provider options
  // most important are: model, weakModel, apiKey, and apiBaseUrl
  llmOptions?: LLMOptions

  // optional, file/folder/glob-specific rule overrides
  overrides?: Array<{
    include?: string[] // array of glob files to apply this override to
    exclude?: string[] // array of glob files to exclude from this override
    rules?: Record<string, 'error' | 'warn' | 'off'>
  }>
}
```

See the [source](https://github.com/gptlint/gptlint/tree/main/src/config.ts) for more details.

## Custom Rules

By default, `ruleFiles` will search `[ '.gptlint/**/*.md' ]` for project-specific rule definition markdown files.

You can override this with the `ruleFiles` config setting or with the `--rules` (`-r`) CLI flag.

### Code-Based Custom Rules

You can also specify code-based custom [RuleDefinition](https://github.com/gptlint/gptlint/tree/main/src/rule.ts) objects which are normal JS/TS objects with hooks for fully customizing linting behavior.

An example of a code-based rule (the equivalent of a plugin in `eslint`) is [effective-tsconfig](https://github.com/gptlint/gptlint/tree/main/rules/custom/effective-tsconfig.ts), which has a `project` `scope` and uses the `preProcessProject` hook to resolve a project's `tsconfig.json` file and check for best practices.

Here's an example:

```js filename="my-custom-rule.js"
/** @type {import('gptlint').RuleDefinition} */
export default {
  name: 'my-custom-rule',
  title: 'Example rule',
  level: 'error',
  scope: 'file',

  preProcessFile: async ({ file, rule, cache, config, cwd }) => {
    return {
      lintErrors: [
        {
          message: 'This is an example custom lint error',
          level: 'error'
        }
      ]
    }
  }
}
```

```js filename="gptlint.config.js"
import { recommendedConfig } from 'gptlint'
import myCustomRule from './my-custom-rule.js'

/** @type {import('gptlint').GPTLintConfig} */
export default [
  ...recommendedConfig,
  {
    ruleDefinitions: [myCustomRule]
  }
]
```

## Example Config Files

**Using `gpt-4-turbo-preview`** as the strong model:

```js filename="gptlint.config.js"
import { recommendedConfig } from 'gptlint'

/** @type {import('gptlint').GPTLintConfig} */
export default [
  ...recommendedConfig,
  {
    llmOptions: {
      model: 'gpt-4-turbo-preview',
      weakModel: 'gpt-3.5-turbo'
    }
  }
]
```

**Using Anthropic Claude via OpenRouter:**

See [Using Anthropic](./llm-providers.md#anthropic) for more info.

```js filename="gptlint.config.js"
import { recommendedConfig } from 'gptlint'

/** @type {import('gptlint').GPTLintConfig} */
export default [
  ...recommendedConfig,
  {
    llmOptions: {
      apiBaseUrl: 'https://openrouter.ai/api/v1',
      model: 'anthropic/claude-3-opus:beta',
      weakModel: 'anthropic/claude-3-haiku:beta',
      // Optional
      kyOptions: {
        headers: {
          // Optional, for including your app on openrouter.ai rankings.
          'HTTP-Referer': 'https://gptlint.dev',
          // Optional, shows in rankings on openrouter.ai.
          'X-Title': 'gptlint'
        }
      }
    }
  }
]
```

## Differences from ESLint Config Format

Currently, all `gptlint` config objects are merged together into a single, flat, final config. This differs from ESLint's config array in that individual ESLint config objects can target different files with different config settings. Here we chose to use an optional `overrides` field so you can customize rule settings for specific files.

<Callout>
  If you're ever not sure how your config is being resolved or why you're seeing
  unexpected behavior, double-check the output of `gptlint --print-config`.
</Callout>
