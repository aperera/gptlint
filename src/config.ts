import type { MergeDeep, SetRequired, Simplify } from 'type-fest'
import type { SimplifyDeep } from 'type-fest/source/merge-deep.js'
import findCacheDirectory from 'find-cache-dir'
import { z } from 'zod'

import { dedupe, getEnv, pruneUndefined } from './utils.js'

export const LinterConfigRuleSettingSchema = z.enum(['off', 'warn', 'error'])
export type LinterConfigRuleSetting = z.infer<
  typeof LinterConfigRuleSettingSchema
>

export const LinterConfigRuleSettingsSchema = z.record(
  z.string(),
  LinterConfigRuleSettingSchema
)
export type LinterConfigRuleSettings = z.infer<
  typeof LinterConfigRuleSettingsSchema
>

// TODO: update when we decide on a project name
export const defaultCacheDir =
  findCacheDirectory({ name: 'gptlint' }) ?? '.gptlint'

export const LLMOptionsSchema = z.object({
  model: z
    .string()
    .optional()
    .describe('Which LLM to use for assessing rule conformance.'),

  weakModel: z
    .string()
    .optional()
    .nullable()
    .describe(
      'If defined, will use a two-pass approach for assessing rule conformance. The `weakModel` should be cheaper and will be used to generate potential rule violations, with the stronger `model` being used in a second pass to validate these potential rule violations and filter out false positives. Set to "none" or `null` to disable two-pass linting.'
    ),

  temperature: z
    .number()
    .min(0.0)
    .max(2.0)
    .optional()
    .describe('LLM temperature parameter.'),

  modelSupportsJsonResponseFormat: z
    .boolean()
    .optional()
    .describe(
      "A Boolean value indicating whether or not `model` supports OpenAI's JSON output mode."
    ),

  apiKey: z
    .string()
    .optional()
    .describe(
      'API key for the OpenAI-compatible LLM API. Defaults to the value of the `OPENAI_API_KEY` environment variable.'
    ),

  apiOrganizationId: z
    .string()
    .optional()
    .describe(
      'Optional organization ID that should be billed for LLM API requests. This is only necessary if your OpenAI API key is scoped to multiple organizations.'
    ),

  apiBaseUrl: z
    .string()
    .optional()
    .describe(
      'Base URL for the OpemAI-compatible LLM API. Defaults to the OpenAI API `https://api.openai.com/v1`'
    ),

  kyOptions: z
    .record(z.any())
    .optional()
    .describe(
      'Additional options for customizing HTTP calls to the LLM API, such as custom `headers` to pass with every request. All options are passed to `ky` which is the HTTP `fetch` wrapper used under the hood.'
    )
})
export type LLMOptions = z.infer<typeof LLMOptionsSchema>

export const LinterOptionsSchema = z.object({
  noInlineConfig: z
    .boolean()
    .optional()
    .describe('A Boolean value for whether inline configuration is allowed.'),

  earlyExit: z
    .boolean()
    .optional()
    .describe(
      'A Boolean value for whether to exit greedily after finding the first error.'
    ),

  debug: z.boolean().optional().describe('Enables debug logging.'),
  debugConfig: z
    .boolean()
    .optional()
    .describe(
      'When enabled, logs the resolved config and parsed rules and then exits.'
    ),
  debugModel: z.boolean().optional().describe('Enables verbose LLM logging.'),
  debugStats: z
    .boolean()
    .optional()
    .describe(
      'Enables logging of cumulative LLM stats (total tokens and cost).'
    ),

  noCache: z.boolean().optional().describe('Disables the built-in cache.'),
  disabled: z.boolean().optional().describe('Disables linting entirely.'),

  cacheDir: z
    .string()
    .optional()
    .describe('A string path to the shared cache directory.'),

  concurrency: z
    .number()
    .int()
    .nonnegative()
    .optional()
    .describe('Limits the maximum number of concurrent tasks.')
})
export type LinterOptions = z.infer<typeof LinterOptionsSchema>

export const LinterConfigSchema = z.object({
  files: z
    .array(z.string())
    .optional()
    .describe(
      'An array of glob patterns for the files to process. If not specified, the configuration object applies to all files matched by any other configuration object.'
    ),

  ignores: z
    .array(z.string())
    .optional()
    .describe('An array of glob patterns for files that should be ignored.'),

  ruleFiles: z
    .array(z.string())
    .optional()
    .describe('An array of glob patterns to rule definition markdown files.'),

  rules: LinterConfigRuleSettingsSchema.optional().describe(
    'An object customizing the configured rules.'
  ),

  linterOptions: LinterOptionsSchema.optional().describe(
    'An object containing settings related to the linting process.'
  ),

  llmOptions: LLMOptionsSchema.optional().describe('')
})
export type LinterConfig = z.infer<typeof LinterConfigSchema>
export type GPTLintConfig = LinterConfig[]

export type ResolvedLinterConfig = Simplify<
  Omit<
    SetRequired<LinterConfig, keyof LinterConfig>,
    'linterOptions' | 'llmOptions'
  > & {
    linterOptions: SetRequired<LinterOptions, keyof LinterOptions>
    llmOptions: SetRequired<LLMOptions, 'model'>
  }
>

export const defaultLinterOptions: Readonly<LinterOptions> = {
  noInlineConfig: false,
  earlyExit: false,
  concurrency: 16,
  debug: false,
  debugConfig: false,
  debugModel: false,
  debugStats: true,
  disabled: false,
  noCache: false,
  cacheDir: defaultCacheDir
}

export const defaultLLMOptions: Readonly<LLMOptions> = {
  apiKey: getEnv('OPENAI_API_KEY'),
  apiBaseUrl: 'https://api.openai.com/v1',
  // The default model is `gpt-4`. We're not using `gpt-4-turbo-preview` as the
  // default because some developers don't have access to it, and we're not
  // using `gpt-3.5-turbo` as the default because it doesn't perform as well in
  // our tests.
  model: 'gpt-4',
  // model: 'gpt-4-turbo-preview',
  weakModel: 'gpt-3.5-turbo',
  temperature: 0
}

export const defaultLinterConfig: Readonly<
  SetRequired<LinterConfig, 'linterOptions' | 'llmOptions'>
> = {
  linterOptions: defaultLinterOptions,
  llmOptions: defaultLLMOptions
}

export function parseLinterConfig(config: Partial<LinterConfig>): LinterConfig {
  return LinterConfigSchema.parse(config)
}

export function isValidModel(
  model?: string | null
): model is NonNullable<string> {
  return !!model && model !== 'none'
}

export function mergeLinterConfigs<
  ConfigTypeA extends LinterConfig = LinterConfig,
  ConfigTypeB extends LinterConfig = LinterConfig
>(
  configA: ConfigTypeA,
  configB: ConfigTypeB
): SimplifyDeep<MergeDeep<ConfigTypeA, ConfigTypeB>> {
  return {
    ...pruneUndefined(configA),
    ...pruneUndefined(configB),
    rules: { ...configA.rules, ...configB.rules },
    ignores:
      configA.ignores || configB.ignores
        ? dedupe([...(configA.ignores ?? []), ...(configB.ignores ?? [])])
        : undefined,
    linterOptions:
      configA.linterOptions || configB.linterOptions
        ? {
            ...pruneUndefined(configA.linterOptions ?? {}),
            ...pruneUndefined(configB.linterOptions ?? {})
          }
        : undefined,
    llmOptions:
      configA.llmOptions || configB.llmOptions
        ? {
            ...pruneUndefined(configA.llmOptions ?? {}),
            ...pruneUndefined(configB.llmOptions ?? {})
          }
        : undefined
  } as any
}
