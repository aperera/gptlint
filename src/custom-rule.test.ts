import { assert, expect, test } from 'vitest'

import type * as types from './types.js'
import customRule from '../rules/prefer-fetch-over-axios.js'
import { parseLinterConfig } from './config.js'

test('custom rule - prefer-fetch-over-axios', async () => {
  assert(customRule.preProcessFile)

  const rawLinterConfig: Partial<types.LinterConfig> = {
    ruleDefinitions: [customRule]
  }

  const linterConfig = parseLinterConfig(rawLinterConfig)
  assert(linterConfig)
  expect(linterConfig.ruleDefinitions?.length).toBe(1)

  const rule = linterConfig.ruleDefinitions?.[0]!
  expect(rule).toBeTruthy()
  expect(rule.preProcessFile).toBeTruthy()
  expect(rule).toMatchSnapshot()

  console.log(rule)
})
