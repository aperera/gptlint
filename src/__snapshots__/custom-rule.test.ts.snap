// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom rule - prefer-fetch-over-axios 1`] = `
{
  "cacheable": true,
  "desc": "The NPM package \`axios\` should be avoided in favor of native \`fetch\`. Now that native \`fetch\` has widespread support, \`axios\` is effectively deprecated and is generally a code smell when encountered.

Convenience wrappers around \`fetch\` such as \`ky\` and \`ofetch\` are encouraged.

Code which doesn't use the \`axios\` module should be ignored.
",
  "eslint": [
    "no-restricted-imports",
  ],
  "fixable": false,
  "languages": [
    "javascript",
    "typescript",
  ],
  "level": "error",
  "message": "Prefer fetch over axios",
  "metadata": {},
  "name": "prefer-fetch-over-axios",
  "negativeExamples": [],
  "positiveExamples": [],
  "preProcessFile": [Function],
  "scope": "file",
  "source": "./prefer-fetch-over-axios.md",
  "tags": [
    "best practices",
  ],
}
`;
